pkg/login/social/connectors/azuread_oauth.go

```diff
@@ -168,6 +178,69 @@ func (s *SocialAzureAD) UserInfo(ctx context.Context, client *http.Client, token
 	return userInfo, nil
 }
 
+func (s *SocialAzureAD) Exchange(ctx context.Context, code string, authOptions ...oauth2.AuthCodeOption) (*oauth2.Token, error) {
+	s.reloadMutex.RLock()
+	defer s.reloadMutex.RUnlock()
+
+	oauthCfg := s.info
+
+	switch oauthCfg.ClientAuthentication {
+	case social.ManagedIdentity:
+		// Generate client assertion
+		clientAssertion, err := s.ManagedIdentityCallback(ctx, *oauthCfg)
+		if err != nil {
+			return nil, err
+		}
+
+		// Set client assertion parameters
+		authOptions = append(authOptions,
+			oauth2.SetAuthURLParam("client_assertion", clientAssertion),
+			oauth2.SetAuthURLParam("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer"),
+		)
+
+	case social.ClientSecretPost:
+		// Default behavior for ClientSecretPost, no additional setup needed
+
+	default:
+		return nil, fmt.Errorf("invalid client authentication method: %s", oauthCfg.ClientAuthentication)
+	}
+
+	// Default token exchange
+	return s.Config.Exchange(ctx, code, authOptions...)
+}
+
+// ManagedIdentityCallback retrieves a token using the managed identity credential of the Azure service.
+func (s *SocialAzureAD) ManagedIdentityCallback(ctx context.Context, oauthCfg social.OAuthInfo) (string, error) {
+	// Validate required fields for Managed Identity authentication
+	if oauthCfg.ManagedIdentityClientID == "" {
+		return "", fmt.Errorf("ManagedIdentityClientID is required for Managed Identity authentication")
+	}
+	if oauthCfg.Audience == "" {
+		return "", fmt.Errorf("audience is required for Managed Identity authentication")
+	}
+	if err := validateAudience(oauthCfg.Audience); err != nil {
+		return "", err
+	}
```

Comment:
This validation should move to https://github.com/grafana/grafana/blob/main/pkg/login/social/connectors/azuread_oauth.go#L192. So we don't have to do it every time but only when we are updating the settings.

 These fields should only be validated it when using `ManagedIdentity`, otherwise we should require them to be empty

Reply:
@kalleep just submitted changes, feel free to look over and submit any feedback you may have!